File: results/2026-04-08_192624/tasks/16-environment-matrix-generator/typescript-bun-opus/generated-code/matrix-generator.ts
#!/usr/bin/env bun

/**
 * Environment Matrix Generator for GitHub Actions
 *
 * Generates a build matrix (strategy.matrix JSON) from a configuration
 * describing OS options, language versions, feature flags, and
 * include/exclude rules.
 */

// --- Type Definitions ---

/** A single dimension in the matrix (e.g., os, node-version, feature-flag) */
interface MatrixDimension {
  [key: string]: string[];
}

/** A single combination to include or exclude */
interface MatrixEntry {
  [key: string]: string | number | boolean;
}

/** Input configuration for the matrix generator */
interface MatrixConfig {
  /** Named dimensions — each key maps to an array of values */
  dimensions: MatrixDimension;
  /** Combinations to explicitly include (added to the Cartesian product) */
  include?: MatrixEntry[];
  /** Combinations to explicitly exclude (removed from the Cartesian product) */
  exclude?: MatrixEntry[];
  /** Maximum number of parallel jobs (optional) */
  "max-parallel"?: number;
  /** Whether to cancel all jobs if one fails (default: true) */
  "fail-fast"?: boolean;
  /** Maximum allowed matrix size — validation guard */
  "max-size"?: number;
}

/** Output format matching GitHub Actions strategy.matrix */
interface MatrixOutput {
  matrix: {
    include: MatrixEntry[];
  };
  "fail-fast"?: boolean;
  "max-parallel"?: number;
}

// --- Constants ---

/** Default maximum matrix size to prevent accidental combinatorial explosion */
const DEFAULT_MAX_SIZE = 256;

// --- Core Logic ---

/**
 * Validates the input configuration and throws descriptive errors.
 */
function validateConfig(config: MatrixConfig): void {
  if (!config) {
    throw new Error("Configuration is required");
  }

  if (!config.dimensions || typeof config.dimensions !== "object") {
    throw new Error("Configuration must include a 'dimensions' object");
  }

  const keys = Object.keys(config.dimensions);
  if (keys.length === 0) {
    throw new Error("At least one dimension is required");
  }

  for (const key of keys) {
    const values = config.dimensions[key];
    if (!Array.isArray(values)) {
      throw new Error(`Dimension '${key}' must be an array of values`);
    }
    if (values.length === 0) {
      throw new Error(`Dimension '${key}' must have at least one value`);
    }
  }

  if (config["max-parallel"] !== undefined) {
    if (
      typeof config["max-parallel"] !== "number" ||
      config["max-parallel"] < 1 ||
      !Number.isInteger(config["max-parallel"])
    ) {
      throw new Error("'max-parallel' must be a positive integer");
    }
  }

  if (config["max-size"] !== undefined) {
    if (
      typeof config["max-size"] !== "number" ||
      config["max-size"] < 1 ||
      !Number.isInteger(config["max-size"])
    ) {
      throw new Error("'max-size' must be a positive integer");
    }
  }
}

/**
 * Computes the Cartesian product of all dimension value arrays.
 * Returns an array of combination objects.
 */
function cartesianProduct(dimensions: MatrixDimension): MatrixEntry[] {
  const keys = Object.keys(dimensions);
  if (keys.length === 0) return [];

  // Start with the first dimension
  let result: MatrixEntry[] = dimensions[keys[0]].map((v) => ({
    [keys[0]]: v,
  }));

  // Multiply in each subsequent dimension
  for (let i = 1; i < keys.length; i++) {
    const key = keys[i];
    const values = dimensions[key];
    const newResult: MatrixEntry[] = [];
    for (const existing of result) {
      for (const val of values) {
        newResult.push({ ...existing, [key]: val });
      }
    }
    result = newResult;
  }

  return result;
}

/**
 * Checks whether a candidate entry matches a pattern.
 * A pattern matches if every key in the pattern exists in
 * the candidate with the same value.
 */
function entryMatches(candidate: MatrixEntry, pattern: MatrixEntry): boolean {
  for (const key of Object.keys(pattern)) {
    if (String(candidate[key]) !== String(pattern[key])) {
      return false;
    }
  }
  return true;
}

/**
 * Checks if two entries are identical across all keys.
 */
function entriesEqual(a: MatrixEntry, b: MatrixEntry): boolean {
  const aKeys = Object.keys(a).sort();
  const bKeys = Object.keys(b).sort();
  if (aKeys.length !== bKeys.length) return false;
  for (let i = 0; i < aKeys.length; i++) {
    if (aKeys[i] !== bKeys[i]) return false;
    if (String(a[aKeys[i]]) !== String(b[bKeys[i]])) return false;
  }
  return true;
}

/**
 * Main generator function: takes a config and produces the matrix output.
 */
function generateMatrix(config: MatrixConfig): MatrixOutput {
  // Validate input
  validateConfig(config);

  const maxSize = config["max-size"] ?? DEFAULT_MAX_SIZE;

  // Step 1: compute Cartesian product of all dimensions
  let combinations = cartesianProduct(config.dimensions);

  // Step 2: apply exclude rules — remove matching combinations
  if (config.exclude && config.exclude.length > 0) {
    combinations = combinations.filter(
      (combo) => !config.exclude!.some((pattern) => entryMatches(combo, pattern))
    );
  }

  // Step 3: apply include rules — add extra combinations (no duplicates)
  if (config.include && config.include.length > 0) {
    for (const inc of config.include) {
      const alreadyExists = combinations.some((combo) =>
        entriesEqual(combo, inc)
      );
      if (!alreadyExists) {
        combinations.push(inc);
      }
    }
  }

  // Step 4: validate matrix size
  if (combinations.length > maxSize) {
    throw new Error(
      `Matrix size ${combinations.length} exceeds maximum allowed size of ${maxSize}`
    );
  }

  if (combinations.length === 0) {
    throw new Error(
      "Matrix is empty after applying exclude rules — no combinations remain"
    );
  }

  // Step 5: build the output
  const output: MatrixOutput = {
    matrix: {
      include: combinations,
    },
  };

  if (config["fail-fast"] !== undefined) {
    output["fail-fast"] = config["fail-fast"];
  }

  if (config["max-parallel"] !== undefined) {
    output["max-parallel"] = config["max-parallel"];
  }

  return output;
}

// --- CLI entry point ---

async function main(): Promise<void> {
  // Accept config from: CLI arg (file path), or stdin
  let configJson: string;

  const args = process.argv.slice(2);

  if (args.length > 0) {
    // Read config from file
    const filePath = args[0];
    try {
      const file = Bun.file(filePath);
      configJson = await file.text();
    } catch (err: any) {
      console.error(`Error reading config file '${filePath}': ${err.message}`);
      process.exit(1);
    }
  } else {
    // Read config from stdin
    configJson = await new Promise<string>((resolve) => {
      let data = "";
      process.stdin.on("data", (chunk: Buffer) => {
        data += chunk.toString();
      });
      process.stdin.on("end", () => resolve(data));
    });
  }

  if (!configJson.trim()) {
    console.error("Error: No configuration provided. Pass a file path or pipe JSON to stdin.");
    process.exit(1);
  }

  let config: MatrixConfig;
  try {
    config = JSON.parse(configJson);
  } catch {
    console.error("Error: Invalid JSON in configuration input");
    process.exit(1);
  }

  try {
    const result = generateMatrix(config);
    console.log(JSON.stringify(result, null, 2));
  } catch (err: any) {
    console.error(`Error: ${err.message}`);
    process.exit(1);
  }
}

main();

// Export for testing
export {
  generateMatrix,
  validateConfig,
  cartesianProduct,
  entryMatches,
  entriesEqual,
  DEFAULT_MAX_SIZE,
};
export type { MatrixConfig, MatrixOutput, MatrixEntry, MatrixDimension };

File: results/2026-04-17_004319/tasks/16-environment-matrix-generator/default-sonnet46-1m-medium/generated-code/matrix_generator.py
# matrix_generator.py — generates GitHub Actions strategy.matrix JSON from config
import json
import sys
from typing import Any

DEFAULT_MAX_SIZE = 256


class MatrixSizeError(Exception):
    """Raised when generated matrix exceeds the configured maximum size."""


def compute_base_size(matrix_dims: dict[str, list]) -> int:
    """Return the Cartesian product count of all list-valued dimensions."""
    result = 1
    found = False
    for v in matrix_dims.values():
        if isinstance(v, list):
            result *= len(v)
            found = True
    return result if found else 0


def validate_config(config: dict) -> None:
    """Raise ValueError with a meaningful message for invalid configs."""
    os_val = config.get("os")
    if os_val is None:
        raise ValueError("Config must include 'os' key")
    if not isinstance(os_val, list):
        raise ValueError("'os' must be a list of OS strings")
    if len(os_val) == 0:
        raise ValueError("'os' must not be empty")

    if "max_parallel" in config:
        mp = config["max_parallel"]
        if not isinstance(mp, int) or mp < 1:
            raise ValueError("'max_parallel' must be a positive integer")

    if "max_size" in config:
        ms = config["max_size"]
        if not isinstance(ms, int) or ms < 1:
            raise ValueError("'max_size' must be a positive integer")


def generate_matrix(config: dict) -> dict:
    """Generate GitHub Actions strategy object from a config dict.

    Config keys:
      os               — list of OS strings (required)
      language_versions— dict mapping dimension name -> list of version strings
      feature_flags    — dict mapping flag name -> list of values (optional)
      include          — list of extra combination dicts (optional)
      exclude          — list of combination dicts to remove (optional)
      fail_fast        — bool (default True)
      max_parallel     — int (optional)
      max_size         — int max Cartesian-product size (default 256)

    Returns the top-level strategy object (fail-fast, max-parallel, matrix).
    Raises MatrixSizeError if base Cartesian product exceeds max_size.
    """
    validate_config(config)

    max_size: int = config.get("max_size", DEFAULT_MAX_SIZE)

    # Build the matrix dimensions dict
    matrix: dict[str, Any] = {"os": config["os"]}

    for dim, versions in config.get("language_versions", {}).items():
        matrix[dim] = versions

    for flag, values in config.get("feature_flags", {}).items():
        matrix[flag] = values

    # Validate size before optional include/exclude which may not grow base
    base_size = compute_base_size(matrix)
    if base_size > max_size:
        total = base_size
        raise MatrixSizeError(
            f"Matrix size {total} exceeds maximum allowed size of {max_size}. "
            f"Reduce dimensions or increase max_size."
        )

    if "include" in config:
        matrix["include"] = config["include"]
    if "exclude" in config:
        matrix["exclude"] = config["exclude"]

    # Build top-level strategy object
    strategy: dict[str, Any] = {
        "fail-fast": config.get("fail_fast", True),
        "matrix": matrix,
    }

    if "max_parallel" in config:
        strategy["max-parallel"] = config["max_parallel"]

    return strategy


def main(argv: list[str] | None = None) -> int:
    """CLI entry point. Reads a JSON config file and prints the matrix JSON."""
    args = sys.argv[1:] if argv is None else argv

    if len(args) != 1:
        print(f"Usage: {sys.argv[0]} <config.json>", file=sys.stderr)
        return 1

    config_path = args[0]
    try:
        with open(config_path) as f:
            config = json.load(f)
    except FileNotFoundError:
        print(f"Error: config file not found: {config_path}", file=sys.stderr)
        return 1
    except json.JSONDecodeError as exc:
        print(f"Error: invalid JSON in config file: {exc}", file=sys.stderr)
        return 1

    try:
        strategy = generate_matrix(config)
    except (ValueError, MatrixSizeError) as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1

    print(json.dumps(strategy, indent=2))
    return 0


if __name__ == "__main__":
    sys.exit(main())
